core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushProjectionIntoTableScan.java

```diff
@@ -98,28 +98,20 @@ public Result apply(ProjectNode project, Captures captures, Context context)
     {
         TableScanNode tableScan = captures.get(TABLE_SCAN);
 
-        Map<Symbol, Expression> inputExpressions = project.getAssignments().getMap();
-
-        ImmutableList.Builder<NodeRef<Expression>> nodeReferencesBuilder = ImmutableList.builder();
-        ImmutableList.Builder<ConnectorExpression> partialProjectionsBuilder = ImmutableList.builder();
-
         // Extract translatable components from projection expressions. Prepare a mapping from these internal
         // expression nodes to corresponding ConnectorExpression translations.
-        for (Map.Entry<Symbol, Expression> expression : inputExpressions.entrySet()) {
-            Map<NodeRef<Expression>, ConnectorExpression> partialTranslations = extractPartialTranslations(
-                    expression.getValue(),
-                    context.getSession(),
-                    typeAnalyzer,
-                    context.getSymbolAllocator().getTypes());
-
-            partialTranslations.forEach((nodeRef, expr) -> {
-                nodeReferencesBuilder.add(nodeRef);
-                partialProjectionsBuilder.add(expr);
-            });
-        }
-
-        List<NodeRef<Expression>> nodesForPartialProjections = nodeReferencesBuilder.build();
-        List<ConnectorExpression> connectorPartialProjections = partialProjectionsBuilder.build();
+        Map<NodeRef<Expression>, ConnectorExpression> partialTranslations = project.getAssignments().getMap().entrySet().stream()
+                .flatMap(expression ->
+                        extractPartialTranslations(
+                                expression.getValue(),
+                                context.getSession(),
+                                typeAnalyzer,
+                                context.getSymbolAllocator().getTypes()).entrySet().stream())
+                // Avoid duplicates
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue, (first, ignore) -> first));
+
+        List<NodeRef<Expression>> nodesForPartialProjections = ImmutableList.copyOf(partialTranslations.keySet());
+        List<ConnectorExpression> connectorPartialProjections = nodesForPartialProjections.stream().map(partialTranslations::get).collect(toImmutableList());
```

Comment:
updated. wanted to avoid relying on ordering properties of `ImmutableMap` methods, as Map interface doesn't seem to enforce that for keySet and values. But seems safe and local enough for readability here.